active/0000-registry-reform.md

```diff
@@ -0,0 +1,109 @@
+- Start Date: 2015-04-09
+- RFC PR:
+- Ember Issue:
+
+# Summary
+
+Fully encapsulate and privatize the `Container` and `Registry` classes by
+exposing a select subset of public methods on `Application` and
+`ApplicationInstance`.
+
+# Motivation
+
+The `Container` and `Registry` classes currently lead a confusing life of
+semi-private exclusion within Ember applications. They are undocumented
+publicly but not fully private either, as knowledge of their particulars is
+required for developing both initializers and unit tests. This situation has
+become untenable as the new `Registry` class has been extracted from
+`Container`, and the complexity of their usage has grown across
+`Application` and `ApplicationInstance` classes.
+
+We can bring sanity to this situation by continuing the work started at the
+`Application` level to expose methods such as `register` and `inject` from the
+internally maintained `Registry`.
+
+Furthermore, once `Container` and `Registry` are fully private, their
+architecture and documentation can be cleaned up. For instance, a
+`Container` can freely reference its associated `Registry` as `registry`
+rather than `_registry`, as it can be assumed that only framework developers
+will reference this property.
+
+# Detailed design
+
+`Application` will expose the following methods from its internally maintained
+registry:
+
+* `register`
+* `inject`
+* `registerOptions` - mapped to `Registry#options`
+* `registerOptionsForType` - mapped to `Registry#optionsForType`
+
+`ApplicationInstance` will also expose the the same methods. However, these
+methods will be exposed from its own internally maintained registry, which
+has the associated `Application`'s registry configured as a "fall back". No
+direct path will be provided from the `ApplicationInstance` to the
+`Application`'s registry.
+
+`ApplicationInstance` will also expose the following methods from its
+internally maintained container:
+
+* `lookup`
+* `lookupFactory`
```

Comment:
Sorry to miss your earlier question. So the `lookup` and `lookupFactory` methods would be exposed directly on the instance, in the same way as `register` and `inject`:

```
App.instanceInitializer({

  initialize: function(instance) {
    instance.lookupFactory('doc-printer');
  }
});
```

(Note: @stefanpenner has concerns about `lookupFactory` as it's currently implemented. It may be replaced entirely for 2.0)